### PR TITLE
Review

### PR DESCRIPTION
**Pull request by koonwen**
I'm gave your library a good look, it's a really cool idea.  Made a PR just so that I have an easy way to do a review and make comments.

A general questions that came to mind:
What's the story for using `ocaml-ambient-context` with Eio code that interoperates with Lwt? From what I understand, the library depends on the specific storage backend to communicate the thread local values. It's not obvious if this will be a problem for code that uses both Eio and Lwt.

### CODE REVIEW

File: lib/ambient_context.ml [hunk 1]
```diff
@@ -27,7 +27,7 @@ let without_binding k cb =
    let (module Store : STORAGE) = Atomic_.get storage in
    Store.without_binding k cb
 
-
+(* Why does this need to be synchronized? *)
 let with_storage_provider new_storage cb : unit =
```
Comment (koonwen) [inline lib/ambient_context.ml:31]: Not that it's bad to be safe, but if I understand the use case correctly, this is the top-level entry point function to install a single storage backend? Not sure why `cas` synchronization is necessary here?

File: lib/ambient_context.ml [hunk 2]
```diff
@@ -1,5 +1,4 @@
 module Atomic_ = Ambient_context_atomic.Atomic
-module TLS = Ambient_context_tls.Thread_local
```
Comment (koonwen) [inline lib/ambient_context.ml:2]: Not sure if this is accidental or on purpose, why would you want to expose this?

File: lib/ambient_context.ml [hunk 3]
```diff
@@ -18,6 +17,7 @@ let get k =
    Store.get k
 
 
+(* Why is bind neccessary to set a value? *)
```
Comment (koonwen) [inline lib/ambient_context.ml:20]: Other TLS's I've seen like the one in `domainslib` has a simple `set` function to add a thread local value, is there a reason why adding the callback is part of the design?

File: README.md [hunk 4]
```diff
@@ -118,7 +118,7 @@ Use `Ambient_context.create_key` to create an opaque key for the value, and expo
 (* lib/foo_deep.ml *)
 module Ctx = Ambient_context
 
-let context_key : string Ctx.key = Ctx.create_key ()
+let header_context_key : string Ctx.key = Ctx.create_key ()
```
[the remark this answers is not in the document]
Comment (ELLIOTTCABLE) [reply README.md:121]: Hm — I wanted to prioritize the instructions for application-developers in the docs, since (theoretically) library-developers will need to _link_ here, if they implement `ambient-context` — I want to make sure that user-flow ("I'm using this Foo Lib, do not care about this 'ambient context' business at all, just tell me what I need to do so I can get back to the task at hand, thx."); but yeah, that means the logical/wallclock order are both backwards in the README 😓

File: lib/ambient_context.ml (hunk 3 above)
Comment (ELLIOTTCABLE) [reply lib/ambient_context.ml:20]: This has actually been a lot of the work behind this — Lwt and Eio both expose a very strange (if you don't know the context) 'stack-delineating' API: "during this stack-frame, asynchronous _binds_ will infect their promises with this additional context."

I have my suspicions about the reasons for that design (apparently it was originally Jeremie Dimino who was responsible for it; but the actual plan is lost to history, afaict) … but at the end of the day, one can be converted to the other trivially …

```ocaml
let with_binding k v cb =
   let old_value = get k in
   set k v;
   cb ();
   set k old_value
```

… but the other cannot be converted to the one; and thus a design that allows for both styles of backend is forced on us.

Nice catch, though! I'm surprised this was noticeable upon a casual reading of the code — perhaps disappointed, even.

File: lib/ambient_context.ml (hunk 1 above)
Comment (ELLIOTTCABLE) [reply lib/ambient_context.ml:32]: So, it's not documented yet, and currently racey, too, until I properly implement copying — but you're _supposed_ to be able to _change_ the storage at runtime.

Rather the point of this whole thing is to have a 'single source of truth' in your binary, that disparate libraries can agree on. The problem is that that source of truth *has* to change: at boot-up, and in some libraries persistently, there is only TLS. But once an asynchronous event-loop starts, TLS becomes unreliable.

(Similarly, although I haven't looked into it in detail yet, there is of course the case of multiple event-loops — I need to look into how `Eio_lwt` actually _plugs_ the Lwt event-loop; I suspect we can just use Eio's storage in the case of both, but I'm not sure? But in any case, it's likely that in a partially-converted Lwt/Eio program, additional “source-of-truth” moves may become necessary at runtime.)

**Comment by ELLIOTTCABLE** [comment]
> A general questions that came to mind:
> What's the story for using ocaml-ambient-context with Eio code that interoperates with Lwt? From what I understand, the library depends on the specific storage backend to communicate the thread local values. It's not obvious if this will be a problem for code that uses both Eio and Lwt.

I haven't looked into it, tbh. As mentioned [above](https://github.com/ELLIOTTCABLE/ocaml-ambient-context/pull/1/files#r1282276697), I'm pretty sure it'll require 'moving' the storage in between event-loop realms (which is already a thing with `Eio_lwt` iirc — okay, I've looked into it _slightly_, lol); but there's already so many dismensions to this problem …

I think this will be a 'write tests first, then make them pass' problem; because it's fairly difficult to reason about. Especially because I don't use Eio myself in the real-world in any meaningful way.

File: lib/ambient_context.ml (hunk 1 above)
Comment (ELLIOTTCABLE) [reply lib/ambient_context.ml:32]: Ah, realized I didn't answer the original question —

well, currently, the configured storage is _not_ thread-local; it's a shared ref, common to all threads. The CAS was necessary to make it safely mutable … but, now that I'm thinking about it, maybe the configured-storage _should_ be, itself, thread-local? Hrm.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ELLIOTTCABLE) [reply README.md:121]: hrm … should prolly use mdx or executable-readme tests or whatever, in any case. hrmph.

[the remark this answers is not in the document]
Comment (koonwen) [reply README.md:121]: Oh, so I was imagining that the more common use case would be application developers that have built up legacy code and want an easy way to inject ad hoc functionality into their deeply nested dependencies. From that perspective, having the instructions flow as "this is how you would hijack the library, then also make sure to add the top level stuff later" seemed more appropriate.

In my mind, `ambient context` was pitched as a last resort so I didn't think other library developers would prioritize implementing it when they could just enforce these "thread-local" values as part of the API interface. For example, passing the http header as a parameter to the API.

That said, you probably meant for `ambient context` as a solution for library maintainers trying to make forward compatible upgrades without breaking it's users. I imagine that unless it's absolutely critical functionality, maintainers would more likely opt out from adding it. If on the other hand, the functionality was really necessary to have, then the library ought to release a new version that breaks users to signal them to upgrade. In which case, now the onus is on the application developers to make a conscious decision to use `ambient context` as a solution for compatibility.

File: lib/ambient_context.ml (hunk 1 above)
Comment (koonwen) [reply lib/ambient_context.ml:32]: > but you're supposed to be able to change the storage at runtime.

Not sure I follow here. My understanding of how this all works is that the storage backend is configured _once_ at runtime prior to the start of the event loop to select the appropriate Eio/Lwt storage provider. So as a consequence, the code is single threaded during the inital mutation.

> The problem is that that source of truth has to change: at boot-up, and in some libraries persistently, there is only TLS. But once an asynchronous event-loop starts, TLS becomes unreliable.

I lose you here about the "source of truth changes in some libraries persistently". What's an example of this?  I was under the impression that the changing "storage backend" is just a wrapper to configure the appropriate concurrency primatives the project uses. Then the management of "thread-local" context is handled by the underlying `Lwt.with_value` function

File: lib/ambient_context.ml (hunk 3 above)
Comment (koonwen) [reply lib/ambient_context.ml:20]: Having thought about this more, I realize that `set` was an api of domain-local storage which has principally different constraints as thread-local storage. Since domains are separate processes with their own isolated memory, they can offer the flexibility of having a direct-style `set` without concern for restoring the previous state.

This is not the case for lightweight-threads because they share the same process memory and therefore the notion of what is "thread-local" is more accurately: _what happened prior in this sequence of binds_. In other words, if you imagine the concurrent program as a directed-acyclic graph with nodes representing branching points. The program can execute and have interleaving branches which can potentially corrupt each other if they retrieve or update the same "thread-local" key. So requiring a callback seems more or less necessary to capture and restore "snapshots" of the thread-local state before executing the next callback

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ELLIOTTCABLE) [reply README.md:121]: Ah, I see the confusion. If you have `app -> bar -> foo`, then this tool is _neither_ about "app author" modifying `foo`, _nor_ about "foo author" being able to modify `foo`'s API — it's _about it being unreasonable for the "bar author" to modify `bar`._

That is, yes, the library-developer you're talking about could totally choose to add a public `~debugging_ctx` argument to every one of their public functions (… which I have personal opinions about, that may or may not be shared by the wider OCaml community … 😓) …

… but to expect _every single intermediate dependency_ to _also_ expose, and pass along, `~debugging_ctx`, is a complete non-starter.

(This is the same overall problem as [W3C's Trace Context](https://www.w3.org/TR/trace-context/) is trying to solve at the "distributed systems" level of the stack; whereas we're solving it in this library in the "dependency tree" layer of the stack.)

File: lib/ambient_context.ml (hunk 1 above)
Comment (ELLIOTTCABLE) [reply lib/ambient_context.ml:32]: > the storage backend is configured once at runtime prior to the start of the event loop to select the appropriate Eio/Lwt storage provider. So as a consequence, the code is single threaded during the inital mutation.

Even if this were true, the event-loop is completely separate from threads — still need to be thread-safe (or in this case, thread-specific), before the event-loop has booted up! It's trivially possible for a program to have forked _before_ it gets to whatever part involves an event-loop (and thus, whatever part might be configuring `ambient-context`.)

> Then the management of "thread-local" context is handled by the underlying Lwt.with_value function

Unfortunately, no — there's three _completely separate_ storage mechanisms; and they have to be converted between at runtime, in edge-cases where "program boot", "some dependency's library initialization", "systhread-fork", and "greenthread event-loop entry" don't happen in the order you want them to:

> What's an example of this?

1. program boots up
2. libraries are initialized, think `let () = Mylib.init …` running
    * a library calls something like `let context_key = Ambient_context.create_key ()`
    * that key now maps to the _thread-local storage_ (a particular and concrete store that is used by default when there's no event-loop)
3. the top-level code reaches a systhread fork
    * the key exists in both threads' separate TLS
4. the top-level code boots up the event loop, and calls `Ambient_context.with_storage_provider …` to configure it
    * oh-oh, now `Mylib.context_key` exists in TLS, _which won't be accessed anymore_
    * when the top-level app — or `Mylib`, even — tries to retrieve it, because now the active storage is, say, LWT, it gets … nothing

tl;dr "we can't synchronize when dependencies and apps attempt to create/access keys; and we can't guarantee it's _after_ a `with_storage_provider` call; even if, in an ideal world, there's only _one_ `with_storage_provider` call."
